### PR TITLE
[SPARK-36642][SQL] Add df.withMetadata: a syntax suger to update the metadata of a dataframe

### PR DESCRIPTION
**Pull request by liangz1**
### What changes were proposed in this pull request?

To make it easy to use/modify the semantic annotation, we want to have a shorter API to update the metadata in a dataframe. Currently we have `df.withColumn("col1", col("col1").alias("col1", metadata=metadata))` to update the metadata without changing the column name, and this is too verbose. We want to have a syntax suger API `df.withMetadata("col1", metadata=metadata)` to achieve the same functionality.

### Why are the changes needed?

A bit of background for the frequency of the update: We are working on inferring the semantic data types and use them in AutoML and store the semantic annotation in the metadata. So in many cases, we will suggest the user update the metadata to correct the wrong inference or add the annotation for weak inference.

### Does this PR introduce _any_ user-facing change?

Yes.
A syntax suger API `df.withMetadata("col1", metadata=metadata)` to achieve the same functionality as`df.withColumn("col1", col("col1").alias("col1", metadata=metadata))`.

### How was this patch tested?

A unit test in DataFrameSuite.scala.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (zero323) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2501]: It would be nice to know the motivation here (maybe we can start with a JIRA ticket?). Adding metadata with `as` (`alias` in Python) is a bit verbose / repetitive but it is not that common operation anyway.

**Comment by HyukjinKwon** [comment]
Okay, had an offline discussion. I am fine with the justification shown in the PR description given that it's going to be used frequently.

**Comment by HyukjinKwon** [comment]
We should add a test BTW.

**Comment by HyukjinKwon** [comment]
Can we file a JIRA and link it in the PR title? I'm neutral on this change. Would need to wait for other committers to review and approve.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:705]: Lets also add a jira prefix in the test title: "SPARK-XXXXX: withMetadata ... "

**Comment by zero323** [comment]
> I'm neutral on this change. 

Same here. The current API is verbose, but metadata always felt more like an internal API (it's not properly documented for starters) and this type of operation can be easily added within the project (for example, with type classes).

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2495]: typo: exisitng -> existing

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2500]: nit: existingColumnName is too verbose - how about `columnName`?

**Comment by liangz1** [comment]
@zero323 Thanks for your comments. I agree that the metadata is not well known to most starters and we may want to improve the documentation for it in a follow-up ticket.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (WeichenXu123) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2500]: shall we use function name `withColumnMetadata` to make it more clear ?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (WeichenXu123) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:710]: nit, add a test which column existing metadata, test replace it with a new metadata

**Review by WeichenXu123 — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (liangz1) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2500]: That would also work, but I prefer the current shorter version, since the first param requires a column name, it's not likely for the users to be confused about whether the metadata is for a column or something else. Also, since this is syntax sugar, a shorter name would better justify the new API.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (liangz1) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:710]: I think that might not be necessary because by default (`df1`), column `"x"` has a metadata `Metadata.empty`, so the new metadata is already "replacing" the existing default metadata.

**Comment by WeichenXu123** [comment]
merged to master

**Pull request closed without merging**
